// react-widgets/lib/CalendarHeader.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Button = _interopRequireDefault(require("./Button"));

var _Icon = require("./Icon");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  label: _propTypes.default.string.isRequired,
  labelId: _propTypes.default.string,
  upDisabled: _propTypes.default.bool.isRequired,
  prevDisabled: _propTypes.default.bool.isRequired,
  todayDisabled: _propTypes.default.bool.isRequired,
  nextDisabled: _propTypes.default.bool.isRequired,
  onViewChange: _propTypes.default.func.isRequired,
  onMoveLeft: _propTypes.default.func.isRequired,
  onMoveToday: _propTypes.default.func.isRequired,
  onMoveRight: _propTypes.default.func.isRequired,
  navigatePrevIcon: _propTypes.default.node,
  navigateNextIcon: _propTypes.default.node,
  localizer: _propTypes.default.object.isRequired,
  isRtl: _propTypes.default.bool
};

function CalendarHeader(_ref) {
  var localizer = _ref.localizer,
      label = _ref.label,
      labelId = _ref.labelId,
      onMoveRight = _ref.onMoveRight,
      onMoveToday = _ref.onMoveToday,
      onMoveLeft = _ref.onMoveLeft,
      onViewChange = _ref.onViewChange,
      prevDisabled = _ref.prevDisabled,
      todayDisabled = _ref.todayDisabled,
      upDisabled = _ref.upDisabled,
      nextDisabled = _ref.nextDisabled,
      _ref$navigatePrevIcon = _ref.navigatePrevIcon,
      navigatePrevIcon = _ref$navigatePrevIcon === void 0 ? _Icon.chevronLeft : _ref$navigatePrevIcon,
      _ref$navigateNextIcon = _ref.navigateNextIcon,
      navigateNextIcon = _ref$navigateNextIcon === void 0 ? _Icon.chevronRight : _ref$navigateNextIcon,
      isRtl = _ref.isRtl;
  return _react.default.createElement("div", {
    className: "rw-calendar-header"
  }, _react.default.createElement(_Button.default, {
    acceptFocus: true,
    id: labelId,
    onClick: onViewChange,
    className: "rw-calendar-btn-view",
    disabled: upDisabled,
    "aria-live": "polite",
    "aria-atomic": "true"
  }, label, " ", _Icon.caretDown), _react.default.createElement(_Button.default, {
    acceptFocus: true,
    className: "rw-calendar-btn-left",
    onClick: onMoveLeft,
    disabled: prevDisabled,
    label: localizer.messages.moveBack(),
    icon: isRtl ? navigateNextIcon : navigatePrevIcon
  }), _react.default.createElement(_Button.default, {
    acceptFocus: true,
    className: "rw-calendar-btn-today",
    disabled: todayDisabled,
    onClick: onMoveToday
  }, localizer.messages.moveToday()), _react.default.createElement(_Button.default, {
    acceptFocus: true,
    className: "rw-calendar-btn-right",
    onClick: onMoveRight,
    disabled: nextDisabled,
    label: localizer.messages.moveForward(),
    icon: isRtl ? navigatePrevIcon : navigateNextIcon
  }));
}

CalendarHeader.propTypes = propTypes;
var _default = CalendarHeader;
exports.default = _default;
module.exports = exports["default"];
