// react-widgets/lib/MultiselectTag.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Button = _interopRequireDefault(require("./Button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var MultiselectTag =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(MultiselectTag, _React$Component);

  function MultiselectTag() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    _this.handleRemove = function (event) {
      var _this$props = _this.props,
          value = _this$props.value,
          disabled = _this$props.disabled,
          onRemove = _this$props.onRemove;
      if (!disabled) onRemove(value, event);
    };

    return _this;
  }

  var _proto = MultiselectTag.prototype;

  _proto.renderDelete = function renderDelete() {
    var _this$props2 = this.props,
        label = _this$props2.label,
        disabled = _this$props2.disabled,
        readOnly = _this$props2.readOnly,
        clearTagIcon = _this$props2.clearTagIcon;
    return _react.default.createElement(_Button.default, {
      variant: null,
      onClick: this.handleRemove,
      className: "rw-multiselect-tag-btn",
      disabled: disabled || readOnly,
      label: label || 'Remove item'
    }, clearTagIcon);
  };

  _proto.render = function render() {
    var _this$props3 = this.props,
        id = _this$props3.id,
        className = _this$props3.className,
        children = _this$props3.children,
        focused = _this$props3.focused,
        disabled = _this$props3.disabled;
    return _react.default.createElement("div", {
      id: id,
      role: "option",
      className: (0, _classnames.default)(className, 'rw-multiselect-tag', disabled && 'rw-state-disabled', focused && !disabled && 'rw-state-focus')
    }, _react.default.createElement("span", {
      className: "rw-multiselect-tag-label"
    }, children), this.renderDelete());
  };

  return MultiselectTag;
}(_react.default.Component);

MultiselectTag.propTypes = {
  id: _propTypes.default.string,
  clearTagIcon: _propTypes.default.node,
  onRemove: _propTypes.default.func.isRequired,
  focused: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  readOnly: _propTypes.default.bool,
  label: _propTypes.default.string,
  value: _propTypes.default.any
};
var _default = MultiselectTag;
exports.default = _default;
module.exports = exports["default"];
